src/controller/CHIPDevice.cpp

```diff
@@ -470,5 +496,29 @@ void Device::AddReportHandler(EndpointId endpoint, ClusterId cluster, AttributeI
     mCallbacksMgr.AddReportCallback(mDeviceId, endpoint, cluster, attribute, onReportCallback);
 }
 
+Device::~Device()
+{
+    if (mExchangeMgr)
+    {
+        // Ensure that any exchange contexts we have open get closed now,
+        // because we don't want them to call back in to us after this
+        // point.
+        mExchangeMgr->CloseAllContextsForDelegate(this);
+    }
+
+    if (mStorageDelegate)
+    {
+        // Store the current device in persistent storage so we have the latest
+        // message counters available next time.
```

Comment:
So we need to do this because:

1. When we deserialize we will deserialize the CASE key information.
2. We don't currently do CASE resumption.

and therefore the fact that there is serialization/deserialization happening here is an implementation detail; it's all one CASE session as far as the spec is concerned, and hence we need message counter continuity.

Is that a correct description of the situation?  If so, it would be good to have the comments (probably at the spot where we deserialize) make this clear.  If not, whatever the reason we need to do this might be, it should be documented in the deserialization code.

Reply:
The CHIPDevice currently has PASE sessions keys which are used for provisioning. The current controller design allows the CHIPDevice to be recreated in the middle of the provisioning flow. This is different from things like OpCreds which are provisioned internally without application input. 

Just as we're ready to provision, the Controller saves the CHIPDevice to storage and notifies the application. Which then recreates the CHIPDevice for network provisioning. 

When it's recreated, the same PASE keys are used to communicate with the device. So for now we've got to make sure we restore the entire state correctly for the PASE session to work, and that's why we need to persist the message counters as well..

Reply:
OK, so the same thing with PASE instead of CASE: doing a serialize/deserialize in what is conceptually (in spec terms) the same PASE session, because we use the same keys and don't redo the handshake.

Reply:
@sagar-apple Can you please add the code comments for this?